solidity/contracts/bank/Bank.sol

```diff
@@ -0,0 +1,299 @@
+        require(msg.sender == address(bridge), "Caller is not the bridge");
+        _;
+    }
+
+    constructor() {
+        cachedChainId = block.chainid;
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    /// @notice Allows the Governance to upgrade the Bridge address.
+    /// @dev The function does not implement any governance delay and does not
+    ///      check the status of the Bridge. The Governance implementation needs
+    ///      to ensure all requirements for the upgrade are satisfied before
+    ///      executing this function.
+    function updateBridge(address _bridge) external onlyOwner {
+        require(_bridge != address(0), "Bridge address must not be 0x0");
+        bridge = _bridge;
+        emit BridgeUpdated(_bridge);
+    }
+
+    /// @notice Moves the given `amount` of balance from the caller to
+    ///         `recipient`.
+    /// @dev Requirements:
+    ///       - `recipient` cannot be the zero address,
+    ///       - the caller must have a balance of at least `amount`.
+    function transferBalance(address recipient, uint256 amount) external {
+        _transferBalance(msg.sender, recipient, amount);
+    }
+
+    /// @notice Sets `amount` as the allowance of `spender` over the caller's
+    ///         balance.
+    /// @dev If the `amount` is set to `type(uint256).max` then
+    ///      `transferBalanceFrom` will not reduce an allowance.
+    ///      Beware that changing an allowance with this function brings the
+    ///      risk that someone may use both the old and the new allowance by
+    ///      unfortunate transaction ordering. One possible solution to mitigate
+    ///      this race condition is to first reduce the spender's allowance to 0
+    ///      and set the desired value afterwards:
+    ///      https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
+    function approveBalance(address spender, uint256 amount) external {
+        _approveBalance(msg.sender, spender, amount);
+    }
+
+    /// @notice Moves `amount` of balance from `spender` to `recipient` using the
+    ///         allowance mechanism. `amount` is then deducted from the caller's
+    ///         allowance unless the allowance was made for `type(uint256).max`.
+    /// @dev Requirements:
+    ///      - `recipient` cannot be the zero address,
+    ///      - `spender` must have a balance of at least `amount`,
+    ///      - the caller must have allowance for `spender`'s balance of at
+    ///        least `amount`.
+    function transferBalanceFrom(
+        address spender,
+        address recipient,
+        uint256 amount
+    ) external {
+        uint256 currentAllowance = allowance[spender][msg.sender];
+        if (currentAllowance != type(uint256).max) {
+            require(
+                currentAllowance >= amount,
+                "Transfer amount exceeds allowance"
+            );
+            _approveBalance(spender, msg.sender, currentAllowance - amount);
```

Comment:
Same comment regarding `unchecked`.

Reply:
Yes, let's do it. It's 162 gas units saving on average for `transferBalanceFrom`.

Reply:
d362e4ea7a506f2cfb0e7e3fe90468eff9275487